celery_amqp_backend/backend.py

```diff
@@ -379,3 +381,259 @@ def __reduce__(self, args=(), kwargs=None):
+        self._cache = kombu.utils.functional.LRUCache(limit=10000)
+
+    def reload_task_result(self, task_id):
+        raise NotImplementedError('reload_task_result is not supported by this backend.')
+
+    def reload_group_result(self, task_id):
+        raise NotImplementedError('reload_group_result is not supported by this backend.')
+
+    def save_group(self, group_id, result):
+        raise NotImplementedError('save_group is not supported by this backend.')
+
+    def restore_group(self, group_id, cache=True):
+        raise NotImplementedError('restore_group is not supported by this backend.')
+
+    def delete_group(self, group_id):
+        raise NotImplementedError('delete_group is not supported by this backend.')
+
+    def add_to_chord(self, chord_id, result):
+        raise NotImplementedError('add_to_chord is not supported by this backend.')
+
+    def get_many(self, task_ids, **kwargs):
+        raise NotImplementedError('get_many is not supported by this backend.')
+
+    def wait_for(self, task_id, timeout=None, interval=0.5, on_interval=None, no_ack=True):
+        """
+        Waits for task and returns the result.
+
+        :param task_id: The task identifiers we want the result for
+        :param timeout: Consumer read timeout
+        :param no_ack: If enabled the messages are automatically acknowledged by the broker
+        :param interval: Interval to drain messages from the queue
+        :param on_interval: Callback function for message poll intervals
+        :param kwargs:
+        :return: Task result body as dict
+        """
+        try:
+            return super().wait_for(
+                task_id,
+                timeout=timeout,
+                interval=interval,
+                no_ack=no_ack,
+                on_interval=on_interval
+            )
+        except exceptions.TimeoutError:
+            consumer = self._consumers.pop(task_id, None)
+            if consumer and consumer not in self._consumers.values():
+                consumer.cancel()
+
+            raise self.WaitTimeoutException()
+
+    def get_task_meta(self, task_id, backlog_limit=1000):
+        def _on_message_callback(message):
+            nonlocal meta, task_id
+            payload = message.decode()
+
+            if not isinstance(payload, (dict,)) or 'task_id' not in payload:
+                return
+
+            if task_id == payload['task_id']:
+                meta = payload
+            else:
+                self._cache[payload['task_id']] = payload
+
+        meta = self._cache.pop(task_id, None)
+
+        if meta is not None:
+            return meta
+
+        consumer = self._consumers.get(task_id)
+
+        if not consumer:
+            return {
+                'status': states.FAILURE,
+                'result': None,
+            }
+
+        consumer.on_message = _on_message_callback
+        consumer.consume()
+
+        try:
+            consumer.connection.drain_events(timeout=0.5)
```

Comment:
This call fails when multiple concurrent executions happen.
Exception: `ConcurrentObjectUseError`
Message: `This socket is already used by another greenlet: <bound method Waiter.switch of <gevent._gevent_c_waiter.Waiter object at 0x7fb6c0c1ebf0>>`
Traceback: 
```
     File \"/usr/local/lib/python3.7/site-packages/celery/result.py\", line 231, in get
    on_message=on_message,
  File \"/usr/local/lib/python3.7/site-packages/celery/backends/base.py\", line 760, in wait_for_pending
    no_ack=no_ack,
  File \"/app/app/celery_amqp_backend/backend.py\", line 471, in wait_for
    on_interval=on_interval
  File \"/usr/local/lib/python3.7/site-packages/celery/backends/base.py\", line 783, in wait_for
    meta = self.get_task_meta(task_id)
  File \"/app/app/celery_amqp_backend/backend.py\", line 510, in get_task_meta
    consumer.connection.drain_events(timeout=0.5)
  File \"/usr/local/lib/python3.7/site-packages/kombu/connection.py\", line 316, in drain_events
    return self.transport.drain_events(self.connection, **kwargs)
  File \"/usr/local/lib/python3.7/site-packages/kombu/transport/pyamqp.py\", line 169, in drain_events
    return connection.drain_events(**kwargs)
  File \"/usr/local/lib/python3.7/site-packages/amqp/connection.py\", line 525, in drain_events
    while not self.blocking_read(timeout):
  File \"/usr/local/lib/python3.7/site-packages/amqp/connection.py\", line 530, in blocking_read
    frame = self.transport.read_frame()
  File \"/usr/local/lib/python3.7/site-packages/amqp/transport.py\", line 294, in read_frame
    frame_header = read(7, True)
  File \"/usr/local/lib/python3.7/site-packages/amqp/transport.py\", line 627, in _read
    s = recv(n - len(rbuf))
  File \"/usr/local/lib/python3.7/site-packages/gevent/_socketcommon.py\", line 663, in recv
    self._wait(self._read_event)
  File \"src/gevent/_hub_primitives.py\", line 317, in gevent._gevent_c_hub_primitives.wait_on_socket
  File \"src/gevent/_hub_primitives.py\", line 322, in gevent._gevent_c_hub_primitives.wait_on_socket
  File \"src/gevent/_hub_primitives.py\", line 297, in gevent._gevent_c_hub_primitives._primitive_wait
```